### PR TITLE
(PA-1991) Load puppet-runtime from a local directory

### PR DESCRIPTION
**Pull request by caseywilliams**
Setting $PUPPET_RUNTIME_PROJECT_PATH to the path of a local clone of
the puppet-runtime repository before building puppet-agent now allows:

- loading initial settings from that project (instead of a fresh clone
  of puppet-runtime based on configs/components/puppet-runtime.json), and
- looking for puppet-runtime tarballs in that project's output directory
  (instead of attempting to download them from Puppet's internal build
  sources)

When this variable is set, the version of puppet-runtime used to build
puppet-agent will be the one currently checked out in
$PUPPET_RUNTIME_PROJECT_PATH (configs/component/puppet-runtime.json
will be ignored); Builds of puppet-agent will fail if no runtime project
for the current puppet-runtime revision and the requested platform has
been built there.

Blocked on https://github.com/puppetlabs/vanagon/pull/563, which adds local settings inheritance.

### CODE REVIEW

File: configs/components/puppet-runtime.rb [hunk 1]
```diff
@@ -1,13 +1,53 @@
...
+    raise "Unable to find a puppet-runtime tarball at '#{tarball_path}'"\
+          " -- Have you built puppet-runtime for the '#{git_describe}' revision and the '#{platform.name}' platform yet?"\
+          unless File.exist?(tarball_path)
+
+    raise "Unable to find a sha1sum file for the puppet-runtime tarball at '#{sha1sum_path}'"\
+          " -- Either rebuild puppet-runtime or ensure this file contains a sha1sum of #{tarball_path}"\
+          unless File.exist?(sha1sum_path)
+
+    pkg.url "file://#{tarball_path}"
+    pkg.sha1sum sha1sum_path
```
Comment (underscorgan) [inline configs/components/puppet-runtime.rb]: This part seems very reasonable / valuable.

File: configs/projects/puppet-agent.rb [hunk 2]
```diff
@@ -98,6 +129,7 @@
   # Provides augeas, curl, libedit, libxml2, libxslt, openssl, puppet-ca-bundle, ruby and rubygem-*
   proj.component "puppet-runtime"
   proj.component "nssm" if platform.is_windows?
+  proj.component "rubygem-puppet-resource_api"
```
[the remark this answers is not in the document]
Comment (caseywilliams) [reply configs/projects/puppet-agent.rb]: Yeah, this should not be here - I'll remove it

File: configs/projects/puppet-agent.rb [hunk 3]
```diff
@@ -1,11 +1,42 @@
...
+  #    process; See puppet-runtime's README.
+  # 4. Set the `PUPPET_RUNTIME_PROJECT_PATH` environment variable to point to
+  #    your local puppet-runtime directory before building puppet-agent. The
+  #    puppet-runtime component will fetch an appropriate puppet-runtime tarball
+  #    from the `output/` directory of that project (all finished builds are
+  #    placed there by default).
+
+  proj.setting(:puppet_runtime_project, 'agent-runtime-5.5.x')
+
+  if ENV['PUPPET_RUNTIME_PROJECT_PATH']
```
[the remark this answers is not in the document]
Comment (caseywilliams) [reply configs/projects/puppet-agent.rb]: This was concerning to others I talked to as well - what do you think about just adding the requirement that the local puppet-runtime clone is pointing to the same revision that's specified in `puppet-runtime.json`? We could fail the build if not.

[the remark this answers is not in the document]
Comment (caseywilliams) [reply configs/projects/puppet-agent.rb]: That's starting to sound a lot like a git submodule, come to think of it

**Comment by caseywilliams** [comment]
I think we'd like to pursue an alternate way of getting these benefits - new PRs coming soon

**Pull request closed without merging**